src/Microsoft.AspNet.Server.Kestrel/Http/Connection.cs

```diff
@@ -63,13 +65,53 @@ public Connection(ListenerContext context, UvStreamHandle socket) : base(context
             ConnectionControl = this;
         }
 
+        private void ExtractIP()
+        {
+            // storage for binary IP addresses
+            Libuv.sockaddr peer;
+            Libuv.sockaddr local;
+            // length of sockaddr structs
+            var peerLength = Marshal.SizeOf(typeof(Libuv.sockaddr));
+            var localLength = peerLength;
+            // buffers for addresses
+            var peerString = new StringBuilder(50);
+            var localString = new StringBuilder(50);
+            // storage for errors during decoding processes
+            Exception peerException;
+            Exception localException;
+
+            // get local and peer IPs in binary form
+            _socket.Libuv.tcp_getsockname((UvTcpHandle)_socket, out local, ref localLength);
+            _socket.Libuv.tcp_getpeername((UvTcpHandle)_socket, out peer, ref peerLength);
+            // decode as IPv4 addresses
+            _socket.Libuv.ip4_name(ref local, localString, 50, out localException);
```

Comment:
interesting. sockaddr on windows has a built in field telling you if it's IPv4 or IPv6. According to this (http://stackoverflow.com/questions/13157151/isnt-struct-sockadr-in-supposed-to-work-for-both-ipv4-and-ipv6) Linux has a similar built in field.

Besides, the rest of your code is hard-coded to IPv4 and will need to be retrofitted for IPv6 later.